contracts/0.4.24/nos/NodeOperatorsRegistry.sol

```diff
@@ -1,702 +1,1169 @@
+        uint256 _nodeOperatorId,
+        uint256 _exitedValidatorsCount,
+        uint256 _stuckValidatorsKeysCount
+    ) external {
+        _onlyExistedNodeOperator(_nodeOperatorId);
+        _auth(STAKING_ROUTER_ROLE);
+
+        _updateExitedValidatorsCount(_nodeOperatorId, uint64(_exitedValidatorsCount), true);
+        _updateStuckValidatorsCount(_nodeOperatorId, uint64(_stuckValidatorsKeysCount));
+    }
+
+    function _updateExitedValidatorsCount(uint256 _nodeOperatorId, uint64 _exitedValidatorsKeysCount, bool _allowDecrease)
+        internal
     {
-        // removing from the last index to the highest one, so we won't get outside the array
-        for (uint256 i = _index.add(_amount); i > _index; --i) {
-            _removeSigningKey(_operator_id, i - 1);
+        Packed64x4.Packed memory signingKeysStats = _loadOperatorSigningKeysStats(_nodeOperatorId);
+        uint64 depositedSigningKeysCount = signingKeysStats.get(DEPOSITED_KEYS_COUNT_OFFSET);
+        uint64 exitedValidatorsCount = signingKeysStats.get(EXITED_KEYS_COUNT_OFFSET);
+
+        if (exitedValidatorsCount != _exitedValidatorsKeysCount) {
+            _requireValidRange(_exitedValidatorsKeysCount <= depositedSigningKeysCount);
+            if (_exitedValidatorsKeysCount < exitedValidatorsCount && !_allowDecrease) {
+                revert("EXITED_VALIDATORS_COUNT_DECREASED");
+            }
+
+            signingKeysStats.set(EXITED_KEYS_COUNT_OFFSET, _exitedValidatorsKeysCount);
+            _saveOperatorSigningKeysStats(_nodeOperatorId, signingKeysStats);
+            emit ExitedSigningKeysCountChanged(_nodeOperatorId, _exitedValidatorsKeysCount);
         }
     }
 
-    /**
-      * @notice Removes a validator signing key #`_index` of operator #`_id` from the set of usable keys. Executed on behalf of Node Operator.
-      * @param _operator_id Node Operator id
-      * @param _index Index of the key, starting with 0
-      */
-    function removeSigningKeyOperatorBH(uint256 _operator_id, uint256 _index) external {
-        require(msg.sender == operators[_operator_id].rewardAddress, "APP_AUTH_FAILED");
-        _removeSigningKey(_operator_id, _index);
+    /// @notice Updates the limit of the validators that can be used for deposit by DAO
+    /// @param _nodeOperatorId Id of the node operator
+    /// @param _targetLimit New number of EXITED validators of the node operator
+    /// @param _isTargetLimitActive active flag
+    function updateTargetValidatorsLimits(uint256 _nodeOperatorId, bool _isTargetLimitActive, uint64 _targetLimit) external {
+        _onlyExistedNodeOperator(_nodeOperatorId);
+        _auth(STAKING_ROUTER_ROLE);
+
+        Packed64x4.Packed memory operatorTargetStats = _loadOperatorTargetValidatorsStats(_nodeOperatorId);
+        operatorTargetStats.set(IS_TARGET_LIMIT_ACTIVE_OFFSET, _isTargetLimitActive ? 1 : 0);
+        operatorTargetStats.set(TARGET_VALIDATORS_COUNT_OFFSET, _isTargetLimitActive ? _targetLimit : 0);
```

Comment:
maybe forbid the following combination inside the code:
```_isTargerLimitActive && _targetLimit == 0``` to avoid human errors and show the intent about 1-based limits storage